### PR TITLE
Added a subclass to store vanilla villager professions

### PR DESCRIPTION
**Pull request by liach**
from https://github.com/CovertJaguar/Railcraft/issues/1152

### CODE REVIEW

File: patches/minecraft/net/minecraft/entity/passive/EntityVillager.java.patch [hunk 1]
```diff
@@ -1,6 +1,14 @@
 --- ../src-base/minecraft/net/minecraft/entity/passive/EntityVillager.java
 +++ ../src-work/minecraft/net/minecraft/entity/passive/EntityVillager.java
-@@ -105,7 +105,7 @@
+@@ -85,6 +85,7 @@
+ import net.minecraft.world.storage.MapData;
+ import net.minecraft.world.storage.MapDecoration;
+ import net.minecraft.world.storage.loot.LootTableList;
++import net.minecraftforge.fml.common.registry.FMLControlledNamespacedRegistry;
```
Comment (ichttt) [inline patches/minecraft/net/minecraft/entity/passive/EntityVillager.java.patch]: No imports in patches

Comment (liach) [reply patches/minecraft/net/minecraft/entity/passive/EntityVillager.java.patch]: Ah idea added it, will change

**Comment by AEnterprise** [comment]
wouldn't it be better to have the villager itself control if it can farm or not (but have it not by default)?

some mods might want to add farming related villagers that do farm

**Comment by LexManos** [comment]
this.getProfessionForge()

**Comment by liach** [comment]
@AEnterprise We can add an event for fixing ai tasks later

**Comment by liach** [comment]
no `this` you mean?

**Comment by LexManos** [comment]
No reason to go through the registry like 4 times just to check if you're a farmer, just use the Forge profession and compare it against farmer -.-

**Comment by liach** [comment]
Hmm, the problem is we need a place to store these profession instances. Can you point out where should I keep the profession instance?

**Comment by williewillus** [comment]
`getProfessionForge().getRegistryName().equals(new ResourceLocation("minecraft:farmer"));` 

(or `.toString().equals("minecraft:farmer")`)

also, you need to patch this in the next method down as well (`onGrowingAdult`)

**Comment by LexManos** [comment]
@ObjectHolder works.

**Comment by Madsthunder** [comment]
Maybe put the `@ObjectHolder` at the head of `VillagerRegistry`, for consistency

https://github.com/MinecraftForge/MinecraftForge/pull/3678/files#diff-bc1c76d98f1c78c885a66259ac290382

**Comment by liach** [comment]
I will make a subclass for it, like all the tests.

**Comment by williewillus** [comment]
The next method in entityvillager also needs a patch, and the fields should be named like `FARMER` (all caps)

**Comment by liach** [comment]
@williewillus 
1. Which next method? 
2. The fields are not uppercased because the registry names are lowercased.

**Comment by williewillus** [comment]
EntityAIHarvestFarmland is added to the villager in another method in this class, so it needs to be patched as well. Don't recall the precise name atm (on mobile) so search for EntityAIHarvestFarmland.

Well, they're constant field names. You remember that we uppercased everything in `Blocks` and `Items` on purpose right? :P

**Comment by liach** [comment]
Aye, will add more annotations.

**Comment by LexManos** [comment]
Thanks for reminding me of #3678 With that change {its pending because he git-derped} this one would not be necessary.

**Comment by liach** [comment]
So merge #3678 first and I change based on that?

**Comment by liach** [comment]
@mezz Can you review?

**Comment by mezz** [comment]
Please don't ask me to review, I will get to it when I can. I am starting with older PRs first.

**Comment by Madsthunder** [comment]
@liach The registry names don't have to be lowercased, whenever the name of the field is passed into a `ResourceLocation`, it is always lowercased.

**Comment by liach** [comment]
Oh thanks @Madsthunder! I thought I should always follow the guideline of the test mods.

**Comment by liach** [comment]
Although there is https://github.com/MinecraftForge/MinecraftForge/commit/93a4e13837cc4139fd33f1f263c71ab8eeb4d82a#diff-e8faa44f7f8d74c6a03b15896167d09cR55, forge stopped writing int professions to nbt, so this is still needed. I expect any vanilla villagers to farm after world reload.

**Comment by LexManos** [comment]
We still write int Profession to nbt

**Comment by mezz** [comment]
Can you deprecate the `FARMER` field from the PR here as discussed at the end of the comments?
https://github.com/MinecraftForge/MinecraftForge/pull/3678#issuecomment-298271219
The `FARMER` added by your PR is in a better organized place and should be the only one.

**Comment by liach** [comment]
Added to my todo list.

**Comment by liach** [comment]
If I recall correctly, there is another bug where villager zombies are named 'entity.Zombie.name'

**Comment by mezz** [comment]
Actually I missed what Lex said, I think this PR is no longer necessary because of the change to `EntityVillager#getProfession` in #3678

**Comment by liach** [comment]
@mezz That patch is in EntityZombieVillager instead of EntityVillager class. Not really.

**Comment by mezz** [comment]
I mean this line changed here:
https://github.com/MinecraftForge/MinecraftForge/pull/3678/files#diff-e8faa44f7f8d74c6a03b15896167d09cR44

**Comment by liach** [comment]
Then how about marking this as a cleanup? It still has its values since it added instance fields for professions and let the patch use the fields, possibly.

**Comment by mezz** [comment]
I'm not sure it's that useful tbh.

**Comment by LexManos** [comment]
Personally I dont think it is. It'd be along the lines of us adding a Professions class to the n.m.init package.. Which I really don't think we should do, people should just use their own @ObjectHolder fields if they are using vanilla profs.

**Comment by mezz** [comment]
Thanks anyway liach, at least the issue is fixed now.

**Comment by liach** [comment]
:smile:

**Pull request closed without merging**